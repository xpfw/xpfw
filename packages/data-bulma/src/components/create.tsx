import * as React from "react"

import { dataOptions,  ICreateHookProps, useCreateWithProps } from "@xpfw/data"
import { getMapToFromProps, iterateSubFields, prependPrefix, SharedField } from "@xpfw/form"
import { get } from "lodash"
import { observer } from "mobx-react-lite"

const BulmaCreate: React.FunctionComponent<ICreateHookProps> = observer((props) => {
  const createProps = useCreateWithProps(props)
  const fields: any[] = []
  iterateSubFields(props.schema, (key, schema) => {
    fields.push(<SharedField key={key} schema={schema} prefix={prependPrefix(getMapToFromProps(props), props.prefix)} />)
  })
  const gotErr = createProps.error != null
  const result = createProps.state
  let msg: any
  if (gotErr) {
    msg = (
      <div className="notification is-danger">
        Error please recheck your inputs or connection {JSON.stringify(createProps.error)}
      </div>
    )
  }
  if (result) {
    msg = (
      <div className="notification is-success">
        Successfully created {get(result, dataOptions.idPath)}
      </div>
    )
  }
  return (
    <div>
      {fields}
      <a className="button is-primary" onClick={createProps.submitCreate}>Create</a>
      {msg}
    </div>
  )

})

export default BulmaCreate
export {
  ICreateHookProps
}
