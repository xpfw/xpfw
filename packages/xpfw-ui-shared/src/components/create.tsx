import { FormErrorStore, FormStore, LoadingStore } from "@xpfw/form-shared"
import { filterFields, globals, IField, IForm, IFormError } from "@xpfw/validate"
import { get, isNil } from "lodash"
import * as React from "react"
import { ComponentBase } from "resub"
import DbStore, { IResult } from "../store/db"
import UserStore from "../store/user"

export interface ISharedFormCreate extends React.Props<any> {
  prefix?: string
  method?: string
  /**
   * Will reset the form values
   */
  resetOnMount?: boolean
  /**
   * Will reset the result of the last creation
   */
  resetState?: boolean
  form: IForm
}

export interface ISharedFormCreateState {
  error?: IFormError
  state?: any
  user?: any
  loading: boolean
}

export type SubmitCreate = () => Promise<any>

export interface IFormCreateProps extends ISharedFormCreate, ISharedFormCreateState {
  submitCreate: SubmitCreate
  /** Prefiltered Fields to display */
  fields: IField[]
}

const submitCreate = (thisRef: {props: ISharedFormCreate}) => {
  return async () => {
    const prefix = get(thisRef, "props.prefix", "")
    await FormStore.validateForm(get(thisRef, "props.form"),
      get(thisRef, "props.method"), prefix, get(thisRef, "state.user"))
    const err =
    FormErrorStore.getError(`${prefix && prefix.length > 0 ? prefix + "." : ""}${get(thisRef, "props.form.model")}`)
    let res
    if (!isNil(err) && err.ok) {
      res = await DbStore.create(get(thisRef, "props.form"), prefix)
    }
    return res
  }
}

function SharedFormCreate<T>(Component: React.ComponentType<IFormCreateProps & T>):
React.ComponentType<ISharedFormCreate & T> {
  return class extends ComponentBase<ISharedFormCreate & T, ISharedFormCreateState> {
    public submitCreate: SubmitCreate
    public constructor(props: ISharedFormCreate & T) {
      super(props)
      this.submitCreate = submitCreate(this)
    }
    public componentWillMount() {
      const prefix: any = get(this.props, "prefix", "")
      const valuePath =
        `${prefix && prefix.length > 0 ? prefix + "." : ""}${get(this.props, "form.model", "undefined")}`
      if (this.props.resetOnMount) {
        FormStore.resetForm(this.props.form, this.props.prefix)
        FormErrorStore.setError(valuePath, undefined)
      }
      if (this.props.resetState) {
        DbStore.setCreateState(valuePath, undefined)
      }
      super.componentWillMount()
    }
    public render() {
      return (
        <Component
          {...this.props}
          {...this.state}
          submitCreate={this.submitCreate}
          fields={filterFields(this.props.form, globals.Method.Create)}
        />
      )
    }
    protected _buildState(props: ISharedFormCreate, initialBuild: boolean): ISharedFormCreateState {
      const prefix = get(props, "prefix", "")
      const valuePath = `${prefix && prefix.length > 0 ? prefix + "." : ""}${get(props, "form.model", "undefined")}`
      return {
        error: FormErrorStore.getError(valuePath),
        state: DbStore.getCreateState(valuePath),
        loading: LoadingStore.getLoading(valuePath),
        user: UserStore.getUser()
      }
    }
  }
}

export default SharedFormCreate
export {
  submitCreate
}
