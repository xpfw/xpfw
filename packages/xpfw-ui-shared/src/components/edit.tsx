import { FormErrorStore, FormStore, LoadingStore } from "@xpfw/form-shared"
import { filterFields, globals, IField, IForm, IFormError } from "@xpfw/validate"
import { get, isNil } from "lodash"
import * as React from "react"
import { ComponentBase } from "resub"
import DbStore from "../store/db"
import UserStore from "../store/user"

export interface ISharedFormEdit extends React.Props<any> {
  prefix?: string
  id?: string
  /**
   * Will reset the form values
   */
  resetOnMount?: boolean
  /**
   * Will reset the result of the last creation
   */
  resetState?: boolean
  form: IForm
}

export interface ISharedFormEditState {
  error?: IFormError
  state?: any
  original?: any
  user?: any
  loading: boolean
}

export type SubmitEdit = () => Promise<any>

export interface IFormEditProps extends ISharedFormEdit, ISharedFormEditState {
  submitEdit: SubmitEdit
  /** Prefiltered Fields to display */
  fields: IField[]
}

const submitEdit = (thisRef: {props: ISharedFormEdit}) => {
  return async () => {
    const prefix = get(thisRef, "props.prefix", "")
    await FormStore.validateForm(get(thisRef, "props.form"),
      globals.Method.Update, prefix, get(thisRef, "state.user"))
    const err =
    FormErrorStore.getError(`${prefix && prefix.length > 0 ? prefix + "." : ""}${get(thisRef, "props.form.model")}`)
    let res
    if (!isNil(err) && err.ok) {
      res = await DbStore.patch(get(thisRef, "props.id"),
        get(thisRef, "props.form"), get(thisRef, "props.prefix"))
    }
    return res
  }
}

function SharedFormEdit<T>(Component: React.ComponentType<IFormEditProps & T>):
React.ComponentType<ISharedFormEdit & T> {
  return class extends ComponentBase<ISharedFormEdit & T, ISharedFormEditState> {
    public submitEdit: SubmitEdit
    public constructor(props: ISharedFormEdit & T) {
      super(props)
      this.submitEdit = submitEdit(this)
    }
    public componentWillMount() {
      const prefix: any = get(this.props, "prefix", "")
      const valuePath = `${prefix && prefix.length > 0 ? prefix + "." : ""}${get(this.props, "id", "")}${get(this.props, "form.model", "undefined")}`
      if (this.props.resetOnMount) {
        FormErrorStore.setError(valuePath, undefined)
      }
      if (this.props.resetState) {
        DbStore.setUpdateState(valuePath, undefined)
      }
      super.componentWillMount()
    }
    public render() {
      return (
        <Component
          {...this.props}
          {...this.state}
          submitEdit={this.submitEdit}
          fields={filterFields(this.props.form, globals.Method.Update)}
        />
      )
    }
    protected _buildState(props: ISharedFormEdit, initialBuild: boolean): ISharedFormEditState {
      const prefix: any = get(props, "prefix", "")
      const valuePath = `${prefix && prefix.length > 0 ? prefix + "." : ""}${get(props, "id", "")}${get(props, "form.model", "undefined")}`
      const id: any = get(props, "id", "none")
      const collection = get(props, "form.collection", "none")
      return {
        error: FormErrorStore.getError(valuePath),
        state: DbStore.getUpdateState(valuePath),
        loading: LoadingStore.getLoading(valuePath),
        original: DbStore.getEditOriginal(id, get(props, "form"), get(props, "prefix", "")),
        user: UserStore.getUser()
      }
    }
  }
}

export default SharedFormEdit
export {
  submitEdit
}
