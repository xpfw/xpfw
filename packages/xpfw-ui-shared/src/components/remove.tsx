import { FormErrorStore, FormStore, LoadingStore } from "@xpfw/form-shared"
import { globals, IForm, IFormError } from "@xpfw/validate"
import { get } from "lodash"
import * as React from "react"
import { ComponentBase } from "resub"
import DbStore, { REMOVE_ADDON_KEY } from "../store/db"

export interface ISharedFormRemove extends React.Props<any> {
  id?: string
  form: IForm
}

export interface ISharedFormRemoveState {
  error?: IFormError
  state?: any
  loading: boolean
}

export type SubmitRemove = () => Promise<any>

export interface IFormRemoveProps extends ISharedFormRemove, ISharedFormRemoveState {
  submitRemove: SubmitRemove
}

const submitRemove = (thisRef: {props: ISharedFormRemove}) => {
  return async () => {
    return DbStore.remove(get(thisRef, "props.id"),
        get(thisRef, "props.form.collection"))
  }
}

function SharedFormRemove<T>(Component: React.ComponentType<IFormRemoveProps & T>):
React.ComponentType<ISharedFormRemove & T> {
  return class extends ComponentBase<ISharedFormRemove & T, ISharedFormRemoveState> {
    public submitRemove: SubmitRemove
    public constructor(props: ISharedFormRemove & T) {
      super(props)
      this.submitRemove = submitRemove(this)
    }
    public render() {
      return (
        <Component
          {...this.props}
          {...this.state}
          submitRemove={this.submitRemove}
        />
      )
    }
    protected _buildState(props: ISharedFormRemove, initialBuild: boolean): ISharedFormRemoveState {
      const id: any = get(props, "id", "_")
      return {
        state: DbStore.getRemoveState(id),
        loading: LoadingStore.getLoading(REMOVE_ADDON_KEY + id)
      }
    }
  }
}

export default SharedFormRemove
export {
  submitRemove
}
