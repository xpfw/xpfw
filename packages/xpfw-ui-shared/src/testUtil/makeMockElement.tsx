import { cloneDeep, get, isArray, isNil, isObject, isString, set } from "lodash"
import * as React from "react"

const idtmp: string[] = []

const makeMockElement = (name: string, stripIdsAt?: string | string[], idPath: string = "_id") => {
  return class extends React.Component<any, any> {
    public render() {
      const toStringify = this.props
      const stripMe = isArray(stripIdsAt) ? stripIdsAt : [stripIdsAt]
      if (isArray(stripIdsAt)) {
        for (const stripper of stripIdsAt) {
          if (isString(stripper)) {
            let toStrip = get(this.props, stripper)
            if (!isArray(toStrip)) {
              toStrip = [toStrip]
            }
            const newIds = []
            for (const obj of toStrip) {
              const id = isString(obj) ? obj : get(obj, idPath)
              let newId = idtmp.indexOf(id)
              if (newId === -1) {
                newId = idtmp.length + 1
                idtmp.push(id)
              }
              const objClone = cloneDeep(obj)
              if (isString(obj)) {
                newIds.push(newId)
              } else {
                set(objClone, idPath, newId)
                newIds.push(objClone)
              }
            }
            if (newIds.length > 0 && !isNil(newIds[0])) {
              set(toStringify, stripper, newIds)
            }
          }
        }
      }
      return (
        <div className={name}>
          {JSON.stringify(toStringify, undefined, 2)}
        </div>
      )
    }
  }
}

export default makeMockElement
