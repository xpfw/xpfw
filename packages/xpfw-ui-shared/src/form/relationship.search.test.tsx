import { ComponentRegistry, FormStore, SharedField } from "@xpfw/form-shared"
import { getRandomApp } from "@xpfw/test-util"
import { FeathersClient } from "@xpfw/ui-feathers"
import { FieldType, TestDefs,  ValidationRegistry } from "@xpfw/validate"
import "isomorphic-fetch"
import * as render from "preact-render-to-json"
import * as React from "react"
import BackendClient from "../client"
import { submitCreate } from "../components/create"
import SharedFormList from "../components/list"
import DbStore from "../store/db"
import makeMockElement from "../testUtil/makeMockElement"
import RelationShipWrapper, { addId, getListFormFromRelationshipField, removeId, searchRelated } from "./relationship"

BackendClient.client = FeathersClient

test("Relationship searching Test", async () => {
  const form = TestDefs.FormNumberAndRequiredText
  const s: any = [TestDefs.FormNumberAndRequiredText.collection, TestDefs.FormRelationshipToNumAndRequiredText.collection]
  const relationCol: any = TestDefs.FormNumberAndRequiredText.collection
  const appRef = await getRandomApp(s, true, BackendClient.client, false, {convertIds: true})
  TestDefs.FormNumberAndRequiredText.options = {idPath: "_id", filterOutById: true}
  TestDefs.FormRelationshipToNumAndRequiredText.options = {idPath: "_id", filterOutById: true}
  ValidationRegistry.registerForm(TestDefs.FormNumberAndRequiredText)
  ValidationRegistry.registerForm(TestDefs.FormRelationshipToNumAndRequiredText)
  const MockEle: any = makeMockElement("show",
    ["list.result", "queryData._id.$nin", "queryData.customPre._id.$nin"])
  const MockedRelationship = RelationShipWrapper(MockEle)
  ComponentRegistry.registerComponent(FieldType.RelationshipSingle, MockedRelationship)
  ComponentRegistry.registerComponent(FieldType.RelationshipMulti, MockedRelationship)
  const createdEles: any = []
  for (let i = 0; i < 10; i++) {
    FormStore.setValue(TestDefs.RequiredTextField.mapTo, "my " + (i % 2 === 0 ? " iText " : "null") + i)
    FormStore.setValue(TestDefs.NumberField.mapTo, 420 + i)
    createdEles.push(await submitCreate({props: {form}})())
  }
  await DbStore.getFromServer(createdEles[1].result.id, relationCol)

  // Multi
  const prefixes = ["", "customPre"]
  for (const prefix of prefixes) {
    expect(render(<SharedField field={TestDefs.RelationshipMultiField} />)).toMatchSnapshot("nothing yet")
    const thisRef: any = {props: {field: TestDefs.RelationshipMultiField, prefix}}
    const boundSearchRelated = searchRelated(thisRef)
    const searchForm = getListFormFromRelationshipField(TestDefs.RelationshipMultiField)
    const MockedList = SharedFormList(MockEle)
    const mySearch = await boundSearchRelated("my")
    expect(render(<SharedField field={TestDefs.RelationshipMultiField} />)).toMatchSnapshot(prefix + "search for my")
    expect(render(<MockedList prefix={prefix} form={searchForm} />)).toMatchSnapshot(prefix + "list search for my")
    await boundSearchRelated("Text")
    expect(render(<SharedField field={TestDefs.RelationshipMultiField} />)).toMatchSnapshot(prefix + "search for Text")
    expect(render(<MockedList prefix={prefix} form={searchForm} />)).toMatchSnapshot(prefix + "list search for Text")
    await boundSearchRelated("null")
    expect(render(<SharedField field={TestDefs.RelationshipMultiField} />)).toMatchSnapshot(prefix + "search for null")
    expect(render(<MockedList prefix={prefix} form={searchForm} />)).toMatchSnapshot(prefix + "list search for null")
    const val = createdEles[0].result._id
    FormStore.setValue(
      `${prefix && prefix.length > 0 ? prefix + "." : ""}${TestDefs.RelationshipMultiField.mapTo}`,
      val)
    thisRef.props.value = val
    const filteredSearch = await boundSearchRelated("my")
    FormStore.setValue(
      `${prefix && prefix.length > 0 ? prefix + "." : ""}${TestDefs.RelationshipMultiField.mapTo}`,
      undefined)
    expect(render(<SharedField field={TestDefs.RelationshipMultiField} />))
      .toMatchSnapshot(prefix + "search for my but one value should be excluded")
    expect(render(<MockedList prefix={prefix} form={searchForm} />))
      .toMatchSnapshot(prefix + "list search for my but one value should be excluded")

  }
  await appRef.cleanUp()
}, 10000)
