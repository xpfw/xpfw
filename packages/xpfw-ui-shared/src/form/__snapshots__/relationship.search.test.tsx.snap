// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Relationship searching Test: customPrelist search for Text 1`] = `
<div
  class="show"
>
  {
  "prefix": "customPre",
  "form": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          },
          {
            "mapTo": "_id",
            "type": 2,
            "validate": {
              "type": 2
            }
          }
        ]
      }
    ]
  },
  "list": {
    "result": [
      {
        "_id": 1,
        "myString": "my  iText 0",
        "myNum": 420
      },
      {
        "_id": 3,
        "myString": "my  iText 2",
        "myNum": 422
      },
      {
        "_id": 5,
        "myString": "my  iText 4",
        "myNum": 424
      },
      {
        "_id": 7,
        "myString": "my  iText 6",
        "myNum": 426
      },
      {
        "_id": 9,
        "myString": "my  iText 8",
        "myNum": 428
      }
    ],
    "total": 5,
    "limit": 10,
    "skip": 0
  },
  "loading": true,
  "currentPage": 0,
  "maxPage": 1,
  "queryData": {
    "customPre": {
      "myString": "Text"
    }
  },
  "showNextPage": false,
  "showPrevPage": false
}
</div>
`;

exports[`Relationship searching Test: customPrelist search for my 1`] = `
<div
  class="show"
>
  {
  "prefix": "customPre",
  "form": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          },
          {
            "mapTo": "_id",
            "type": 2,
            "validate": {
              "type": 2
            }
          }
        ]
      }
    ]
  },
  "list": {
    "result": [
      {
        "_id": 1,
        "myString": "my  iText 0",
        "myNum": 420
      },
      {
        "_id": 2,
        "myString": "my null1",
        "myNum": 421
      },
      {
        "_id": 3,
        "myString": "my  iText 2",
        "myNum": 422
      },
      {
        "_id": 4,
        "myString": "my null3",
        "myNum": 423
      },
      {
        "_id": 5,
        "myString": "my  iText 4",
        "myNum": 424
      },
      {
        "_id": 6,
        "myString": "my null5",
        "myNum": 425
      },
      {
        "_id": 7,
        "myString": "my  iText 6",
        "myNum": 426
      },
      {
        "_id": 8,
        "myString": "my null7",
        "myNum": 427
      },
      {
        "_id": 9,
        "myString": "my  iText 8",
        "myNum": 428
      },
      {
        "_id": 10,
        "myString": "my null9",
        "myNum": 429
      }
    ],
    "total": 10,
    "limit": 10,
    "skip": 0
  },
  "loading": true,
  "currentPage": 0,
  "maxPage": 1,
  "queryData": {
    "customPre": {
      "myString": "my"
    }
  },
  "showNextPage": false,
  "showPrevPage": false
}
</div>
`;

exports[`Relationship searching Test: customPrelist search for my but one value should be excluded 1`] = `
<div
  class="show"
>
  {
  "prefix": "customPre",
  "form": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          },
          {
            "mapTo": "_id",
            "type": 2,
            "validate": {
              "type": 2
            }
          }
        ]
      }
    ]
  },
  "list": {
    "result": [
      {
        "_id": 2,
        "myString": "my null1",
        "myNum": 421
      },
      {
        "_id": 3,
        "myString": "my  iText 2",
        "myNum": 422
      },
      {
        "_id": 4,
        "myString": "my null3",
        "myNum": 423
      },
      {
        "_id": 5,
        "myString": "my  iText 4",
        "myNum": 424
      },
      {
        "_id": 6,
        "myString": "my null5",
        "myNum": 425
      },
      {
        "_id": 7,
        "myString": "my  iText 6",
        "myNum": 426
      },
      {
        "_id": 8,
        "myString": "my null7",
        "myNum": 427
      },
      {
        "_id": 9,
        "myString": "my  iText 8",
        "myNum": 428
      },
      {
        "_id": 10,
        "myString": "my null9",
        "myNum": 429
      }
    ],
    "total": 9,
    "limit": 10,
    "skip": 0
  },
  "loading": true,
  "currentPage": 0,
  "maxPage": 1,
  "queryData": {
    "customPre": {
      "myString": "my",
      "_id": {
        "$nin": [
          1
        ]
      }
    }
  },
  "showNextPage": false,
  "showPrevPage": false
}
</div>
`;

exports[`Relationship searching Test: customPrelist search for null 1`] = `
<div
  class="show"
>
  {
  "prefix": "customPre",
  "form": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          },
          {
            "mapTo": "_id",
            "type": 2,
            "validate": {
              "type": 2
            }
          }
        ]
      }
    ]
  },
  "list": {
    "result": [
      {
        "_id": 2,
        "myString": "my null1",
        "myNum": 421
      },
      {
        "_id": 4,
        "myString": "my null3",
        "myNum": 423
      },
      {
        "_id": 6,
        "myString": "my null5",
        "myNum": 425
      },
      {
        "_id": 8,
        "myString": "my null7",
        "myNum": 427
      },
      {
        "_id": 10,
        "myString": "my null9",
        "myNum": 429
      }
    ],
    "total": 5,
    "limit": 10,
    "skip": 0
  },
  "loading": true,
  "currentPage": 0,
  "maxPage": 1,
  "queryData": {
    "customPre": {
      "myString": "null"
    }
  },
  "showNextPage": false,
  "showPrevPage": false
}
</div>
`;

exports[`Relationship searching Test: customPresearch for Text 1`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedByMulti",
    "type": 14,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "searchForm": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          },
          {
            "mapTo": "_id",
            "type": 2,
            "validate": {
              "type": 2
            }
          }
        ]
      }
    ]
  }
}
</div>
`;

exports[`Relationship searching Test: customPresearch for my 1`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedByMulti",
    "type": 14,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "searchForm": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          },
          {
            "mapTo": "_id",
            "type": 2,
            "validate": {
              "type": 2
            }
          }
        ]
      }
    ]
  }
}
</div>
`;

exports[`Relationship searching Test: customPresearch for my but one value should be excluded 1`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedByMulti",
    "type": 14,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "searchForm": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          },
          {
            "mapTo": "_id",
            "type": 2,
            "validate": {
              "type": 2
            }
          }
        ]
      }
    ]
  }
}
</div>
`;

exports[`Relationship searching Test: customPresearch for null 1`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedByMulti",
    "type": 14,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "searchForm": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          },
          {
            "mapTo": "_id",
            "type": 2,
            "validate": {
              "type": 2
            }
          }
        ]
      }
    ]
  }
}
</div>
`;

exports[`Relationship searching Test: list search for Text 1`] = `
<div
  class="show"
>
  {
  "prefix": "",
  "form": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          },
          {
            "mapTo": "_id",
            "type": 2,
            "validate": {
              "type": 2
            }
          }
        ]
      }
    ]
  },
  "list": {
    "result": [
      {
        "_id": 1,
        "myString": "my  iText 0",
        "myNum": 420
      },
      {
        "_id": 3,
        "myString": "my  iText 2",
        "myNum": 422
      },
      {
        "_id": 5,
        "myString": "my  iText 4",
        "myNum": 424
      },
      {
        "_id": 7,
        "myString": "my  iText 6",
        "myNum": 426
      },
      {
        "_id": 9,
        "myString": "my  iText 8",
        "myNum": 428
      }
    ],
    "total": 5,
    "limit": 10,
    "skip": 0
  },
  "loading": true,
  "currentPage": 0,
  "maxPage": 1,
  "queryData": {
    "myString": "Text"
  },
  "showNextPage": false,
  "showPrevPage": false
}
</div>
`;

exports[`Relationship searching Test: list search for my 1`] = `
<div
  class="show"
>
  {
  "prefix": "",
  "form": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          },
          {
            "mapTo": "_id",
            "type": 2,
            "validate": {
              "type": 2
            }
          }
        ]
      }
    ]
  },
  "list": {
    "result": [
      {
        "_id": 2,
        "myString": "my  iText 0",
        "myNum": 420
      },
      {
        "_id": 3,
        "myString": "my null1",
        "myNum": 421
      },
      {
        "_id": 4,
        "myString": "my  iText 2",
        "myNum": 422
      },
      {
        "_id": 5,
        "myString": "my null3",
        "myNum": 423
      },
      {
        "_id": 6,
        "myString": "my  iText 4",
        "myNum": 424
      },
      {
        "_id": 7,
        "myString": "my null5",
        "myNum": 425
      },
      {
        "_id": 8,
        "myString": "my  iText 6",
        "myNum": 426
      },
      {
        "_id": 9,
        "myString": "my null7",
        "myNum": 427
      },
      {
        "_id": 10,
        "myString": "my  iText 8",
        "myNum": 428
      },
      {
        "_id": 11,
        "myString": "my null9",
        "myNum": 429
      }
    ],
    "total": 10,
    "limit": 10,
    "skip": 0
  },
  "loading": true,
  "currentPage": 0,
  "maxPage": 1,
  "queryData": {
    "myString": "my"
  },
  "showNextPage": false,
  "showPrevPage": false
}
</div>
`;

exports[`Relationship searching Test: list search for my but one value should be excluded 1`] = `
<div
  class="show"
>
  {
  "prefix": "",
  "form": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          },
          {
            "mapTo": "_id",
            "type": 2,
            "validate": {
              "type": 2
            }
          }
        ]
      }
    ]
  },
  "list": {
    "result": [
      {
        "_id": 2,
        "myString": "my null1",
        "myNum": 421
      },
      {
        "_id": 3,
        "myString": "my  iText 2",
        "myNum": 422
      },
      {
        "_id": 4,
        "myString": "my null3",
        "myNum": 423
      },
      {
        "_id": 5,
        "myString": "my  iText 4",
        "myNum": 424
      },
      {
        "_id": 6,
        "myString": "my null5",
        "myNum": 425
      },
      {
        "_id": 7,
        "myString": "my  iText 6",
        "myNum": 426
      },
      {
        "_id": 8,
        "myString": "my null7",
        "myNum": 427
      },
      {
        "_id": 9,
        "myString": "my  iText 8",
        "myNum": 428
      },
      {
        "_id": 10,
        "myString": "my null9",
        "myNum": 429
      }
    ],
    "total": 9,
    "limit": 10,
    "skip": 0
  },
  "loading": true,
  "currentPage": 0,
  "maxPage": 1,
  "queryData": {
    "myString": "my",
    "_id": {
      "$nin": [
        1
      ]
    }
  },
  "showNextPage": false,
  "showPrevPage": false
}
</div>
`;

exports[`Relationship searching Test: list search for null 1`] = `
<div
  class="show"
>
  {
  "prefix": "",
  "form": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          },
          {
            "mapTo": "_id",
            "type": 2,
            "validate": {
              "type": 2
            }
          }
        ]
      }
    ]
  },
  "list": {
    "result": [
      {
        "_id": 2,
        "myString": "my null1",
        "myNum": 421
      },
      {
        "_id": 4,
        "myString": "my null3",
        "myNum": 423
      },
      {
        "_id": 6,
        "myString": "my null5",
        "myNum": 425
      },
      {
        "_id": 8,
        "myString": "my null7",
        "myNum": 427
      },
      {
        "_id": 10,
        "myString": "my null9",
        "myNum": 429
      }
    ],
    "total": 5,
    "limit": 10,
    "skip": 0
  },
  "loading": true,
  "currentPage": 0,
  "maxPage": 1,
  "queryData": {
    "myString": "null"
  },
  "showNextPage": false,
  "showPrevPage": false
}
</div>
`;

exports[`Relationship searching Test: nothing yet 1`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedByMulti",
    "type": 14,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "searchForm": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          },
          {
            "mapTo": "_id",
            "type": 2,
            "validate": {
              "type": 2
            }
          }
        ]
      }
    ]
  }
}
</div>
`;

exports[`Relationship searching Test: nothing yet 2`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedByMulti",
    "type": 14,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "searchForm": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          },
          {
            "mapTo": "_id",
            "type": 2,
            "validate": {
              "type": 2
            }
          }
        ]
      }
    ]
  }
}
</div>
`;

exports[`Relationship searching Test: search for Text 1`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedByMulti",
    "type": 14,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "searchForm": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          },
          {
            "mapTo": "_id",
            "type": 2,
            "validate": {
              "type": 2
            }
          }
        ]
      }
    ]
  }
}
</div>
`;

exports[`Relationship searching Test: search for my 1`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedByMulti",
    "type": 14,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "searchForm": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          },
          {
            "mapTo": "_id",
            "type": 2,
            "validate": {
              "type": 2
            }
          }
        ]
      }
    ]
  }
}
</div>
`;

exports[`Relationship searching Test: search for my but one value should be excluded 1`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedByMulti",
    "type": 14,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "searchForm": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          },
          {
            "mapTo": "_id",
            "type": 2,
            "validate": {
              "type": 2
            }
          }
        ]
      }
    ]
  }
}
</div>
`;

exports[`Relationship searching Test: search for null 1`] = `
<div
  class="show"
>
  {
  "field": {
    "mapTo": "ownedByMulti",
    "type": 14,
    "validate": {
      "relationshipCollection": "simpleTestCol",
      "relationshipNamePath": "myString"
    }
  },
  "searchForm": {
    "model": "searchFor.ownedByMulti",
    "collection": "simpleTestCol",
    "sections": [
      {
        "fields": [
          {
            "mapTo": "myString",
            "type": 2,
            "validate": {}
          },
          {
            "mapTo": "_id",
            "type": 2,
            "validate": {
              "type": 2
            }
          }
        ]
      }
    ]
  }
}
</div>
`;
